Update command injection docs (#741)
returntocorp/semgrep-docs: docs/cheat-sheets/java-code-injection.mdx

import CodeInjectionIntro from "/src/components/concept/_code-injection-intro.mdx"

# Code injection prevention for Java

<CodeInjectionIntro />

## 1. Unsafe Reflection

### 1.A. Using reflection with Class.forName

`Class.forName` returns the Class object associated with the class or interface with the given string name. If an attacker can supply values that the application then uses to determine which class to instantiate or which method to invoke, the potential exists for the attacker to create control flow paths through the application that were not intended by the application developers. This attack vector may allow the attacker to bypass authentication or access control checks or otherwise cause the application to behave in an unexpected manner. 
Example:

```java
String userInput = "org.pwned.package"; // value supplied by user input
Class<?> loadClass = Class.forName(userInput + ".MyThread");
```

#### References

- [`Class.forName` documentation](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#forName-java.lang.String-)

#### Mitigation

- Try to avoid non-literal values for the `className` parameter in `Class.forName`.

> - If it is not possible, use a white list for inputs.

Author: ```suggestion
- If it is not possible, use an allow list for inputs.
```
